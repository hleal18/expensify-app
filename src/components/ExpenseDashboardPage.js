import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        This is from my dashboard componente
    </div>
);

export default ExpenseDashboardPage;