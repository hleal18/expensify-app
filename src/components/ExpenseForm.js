import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// Se recibe una instancia de moment.
// tiene el momento actual.
// const now = moment();
// imprime el objeto
// console.log(now);
// imprime la fecha en formato
// console.log(now.format());
// imprime el mes recortado.
// console.log(now.format('MMM'));
// imprime el mes recortado con el dia del mes.
// console.log(now.format('MMM Do'));
// imprime el mes recortado con el dia del mes y el año en cuatro digitos.
// console.log(now.format('MMM Do, YYYY'));


//Se busca aprovechar el estado interno del componente de los datos de los inputs
//y cuando se haya finalizado, esos datos del estado, se guardarán en redux mediante
//el nuevo expense.
class ExpenseForm extends Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    
    //Se enlaza la descripción del estado con la que escribe el usuario.
    //se mantiene guardado el estado en todo momento.
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    //Se busca que para el input de amount, se acepten numeros de hasta dos decimales
    //nada más.
    onAmountChange = (e) => {
        const amount = e.target.value;
        //Se usa una expresión regular. (regular expressiones)
        //es un tema grande a consultar.
        //Se busca la documentación en regex101.com
        //se fuerza que haya por lo menos un numero antes del punto decimal.
        //tambien se permite que el usuario borre toda la cantidad.
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    };
    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            //Set error state equal to 'Please provide description and amount.'
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            //Clear the error
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                note: this.state.note,
                //Se le pasa el timestamp. No el objeto moment.
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add expense</button>
                </form>
            </div>
        );
    }
}

//Una vez terminada la implementación del componente y sus estados queda incluir la conexión
//con los datos de redux.
//Debido a que este componente será reutilizado, no combiene usarlo aquí sino en los componentes
//que se beneficiarán de éste.

export default ExpenseForm;
