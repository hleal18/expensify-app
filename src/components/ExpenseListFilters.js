import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';

//Se encarga de que el input modifique el store y se actualice el filtro
//de forma que apenas se escriba o borre algo, la lista de items se actualice.
const ExpenseListFilters = (props) => (
    <div>
        <input type="text" value={props.filters.text} onChange={(e) => {
            props.dispatch(setTextFilter(e.target.value));
        }} />
    </div>
);

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);
