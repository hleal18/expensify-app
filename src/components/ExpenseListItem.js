import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, id, amount, createdAt, dispatch }) => {
    return (
        <div>
            <Link to={"/edit/" + id}>
                <h3>{description}</h3>
            </Link>
            <p>{amount} - {createdAt}</p>            
        </div>
    );
};

//Cuando no se necesita acceso al state y solo al dispatch, es suficiente no incluir el
//primer par√°metro.
export default ExpenseListItem;
