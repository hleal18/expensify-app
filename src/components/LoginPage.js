import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export class LoginPage extends React.Component {
    onGoogleStartLogin = () => {
        this.props.startLogin('google');
    };
    onGithubStartLogin = () => {
        this.props.startLogin('github');
    };
    render(){
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Expensify</h1>
                    <p>It's time to get your expenses under control.</p>
                    <button className="button" onClick={this.onGoogleStartLogin}>Login with google</button>
                    <button className="button" onClick={this.onGithubStartLogin}>Login with github</button>
                </div>                
            </div>
        );
    }
};

//Se necesita despachar una acciÃ³n.
const mapDispatchToProps = (dispatch) => ({
    startLogin: (providerName) => dispatch(startLogin(providerName))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);

