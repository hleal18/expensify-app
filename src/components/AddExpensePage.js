import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

//Se reutiliza ExpenseForm y se conecta con connect de redux
//Se pasa a una clase
//Como se quiere testear es la versión aislada de AddExpensePage
//Se exporta individualmente (no como default)
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //Debido a la dificultad de testear lo siguiente.
        //Ya que se invoca una función y dentro de ella hay otra.
        //Se implementa el mapDispatchToProps.
        //props.dispatch(addExpense(expense));
        //Se agrega el prop que está enlazado con la nueva
        //funcion.
        this.props.onSubmit(expense);
        //Lo provee router. Permite redireccionar cuando se suban los datos.
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
            <h1>Add expense</h1>
            <ExpenseForm
                onSubmit={this.onSubmit}
            />
            </div>
        );
    }
}

//Se agrega el mapDispatchToProps
//Se ponen las propiedades que invocarán a dispatch.
//Establece la invocación del dispatch, ahora accesible
//desde las props del componente.
const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);