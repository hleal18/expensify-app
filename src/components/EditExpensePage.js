import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onClick = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                {/*Se accede al id que es pasado por el router y accedido mediante
            match y params. Se encuentra en la url.
            Consultar en la consola el history en la herramienta React.
        */}
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onClick}
                >
                    Remove
            </button>
            </div>
        )
    }
}

//Se debe acceder a los props del componente original, con ello es suficiente
//agregar el segundo argumento de props.
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
    }
}

const mapDispathToProps = (dispatch) => ({
    editExpense: (id, expense) => (dispatch(editExpense(id, expense))),
    removeExpense: (data) => (dispatch(removeExpense(data)))
});

export default connect(mapStateToProps, mapDispathToProps)(EditExpensePage);