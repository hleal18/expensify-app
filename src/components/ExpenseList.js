import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => (<ExpenseListItem {...expense} key={expense.id} />))
            )
        }        
    </div>
);

//Componente que se conecta con el store.
//El primer atributo, es una función que recibe el state de la app que
//se pretende usar dentro de la implementación de forma que se defina a que es lo
//que el componente tendrá acceso, en este caso, como ejemplo, se tiene acceso a name.
//El otro parámetro, es el componente al cual se le fusionará el campo del state solicitado
//en su propio campo de props, de forma que en la propia implementación del ExpenseList
//se tiene acceso a props.name que idealmente proviene del Store de redux al cual se
//accedió mediante el método connect.
//Se define el mapStateToProps, que permite mezclar el el campo específico del estado
//con los props del componente que posteriormente se manipula.
//Cada vez que el state cambia, la función de este componente, mapStateToProps es ejecutada.
//ocasionando un re-renderizado.
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//El segundo parámetro se entiene como el parámetro de una función más interna.
//Se retorna un HOC conectado con el state.
//El primer parámetro ocasiona que se suscriba al Store, estando así al asecho de los cambios
//que sucedan para así ejecutarse.
export default connect(mapStateToProps)(ExpenseList);