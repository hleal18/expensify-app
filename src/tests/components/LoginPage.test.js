import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

let wrapper, startLogin;

beforeEach(() => {
    startLogin = jest.fn();
    wrapper = shallow(<LoginPage startLogin={startLogin}/>)
});

//Evalua que el LoginPage se renderice correctamente 
//con respecto al snapshot guardado
test('should render correctly the LoginPage component', () => {
    expect(wrapper).toMatchSnapshot();
});

//Evalua que el metodo startLogin se invoque para el login con google.
test('should call startLogin on button click with google mehod', () => {
    wrapper.find('button').at(0).simulate('click');
    expect(startLogin).toHaveBeenLastCalledWith('google');
});

//Evalua que el metodo startLogin se invoque para el login con github
test('should call startLogin on button click with github method', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(startLogin).toHaveBeenLastCalledWith('github');
})