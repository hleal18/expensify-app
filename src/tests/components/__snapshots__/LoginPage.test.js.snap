// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly the LoginPage component 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p>
      It's time to get your expenses under control.
    </p>
    <div>
      <button
        className="button button--first"
        onClick={[Function]}
      >
        Login with google
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Login with github
      </button>
    </div>
  </div>
</div>
`;
